### PR TITLE
doc: Update readme and add filter docs

### PR DESCRIPTION
**Pull request by VoVAllen**
Signed-off-by: Jinjing.Zhou <allenzhou@tensorchord.ai>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,23 +14,21 @@ pgvecto.rs is a Postgres extension that provides vector similarity search functi
 ## Why use pgvecto.rs
 
 - 💃 **Easy to use**: pgvecto.rs is a Postgres extension, which means that you can use it directly within your existing database. This makes it easy to integrate into your existing workflows and applications.
+- 🔗 **Blazing fast indexing**: pgvecto.rs is fast not only at index creation, but also at insertion with an index. With async creation support, enjoy your search immediately.
 - 🥅 **Filtering**: pgvecto.rs supports filtering. You can set conditions when searching or retrieving points. This is the missing feature of other postgres extensions.
-- 🚀 **High Performance**: pgvecto.rs is designed to provide significant improvements compared to existing Postgres extensions. Benchmarks have shown that its HNSW index can deliver search performance up to 20 times faster than other indexes like ivfflat.
-- 🔧 **Extensible**: pgvecto.rs is designed to be extensible. It is easy to add new index structures and search algorithms. This flexibility ensures that pgvecto.rs can adapt to emerging vector search algorithms and meet diverse performance needs.
+- 🧮 **Quantization**: pgvecto.rs supports scalar quantization and product qutization up to 16x.
```
Comment (usamoi) [inline README.md:19]: 64x

File: README.md [hunk 2]
```diff
@@ -14,23 +14,21 @@ pgvecto.rs is a Postgres extension that provides vector similarity search functi
 ## Why use pgvecto.rs
 
 - 💃 **Easy to use**: pgvecto.rs is a Postgres extension, which means that you can use it directly within your existing database. This makes it easy to integrate into your existing workflows and applications.
+- 🔗 **Blazing fast indexing**: pgvecto.rs is fast not only at index creation, but also at insertion with an index. With async creation support, enjoy your search immediately.
```
Comment (usamoi) [inline README.md:17]: I think "fast indexing" and "async indexing" is different...

File: README.md [hunk 3]
```diff
@@ -14,23 +14,21 @@ pgvecto.rs is a Postgres extension that provides vector similarity search functi
...
-| Parallel Index build                        | ⚡️ Linearly faster with more cores | 🐌 Only single core used   |
-| Index Persistence                           | mmap file                         | Postgres internal storage |
-| WAL amplification                           | 2x 😃                              | 30x 🧐                     |
+|                                             | pgvecto.rs                          | pgvector                 |
+| ------------------------------------------- | ----------------------------------- | ------------------------ |
+| Transaction support                         | ✅                                  | ⚠️                       |
+| Sufficient Result with Delete/Update/Filter | ✅                                  | ⚠️                       |
+| Vector Dimension Limit                      | 65535                               | 2000                     |
+| Prefilter on HNSW                           | ✅                                  | ❌                       |
+| Parallel ans Async Index build              | ⚡️ Linearly faster with more cores | 🐌 Only single core used |
```
Comment (usamoi) [inline README.md:30]: typo here?

File: README.md [hunk 4]
```diff
@@ -39,7 +37,7 @@ And based on our benchmark, pgvecto.rs can be up to 2x faster than pgvector on h
 We recommend you to try pgvecto.rs using our pre-built docker, by running
 
 ```bash
-docker run --name pgvecto-rs-demo -e POSTGRES_PASSWORD=mysecretpassword -p 5432:5432 -d tensorchord/pgvecto-rs:latest
+docker run --name pgvecto-rs-demo -e POSTGRES_PASSWORD=mysecretpassword -p 5432:5432 -d tensorchord/pgvecto-rs:pg16-latest
```
Comment (usamoi) [inline README.md:40]:
```
docker: Error response from daemon: manifest for tensorchord/pgvecto-rs:pg16-latest not found: manifest unknown: manifest unknown.
```

File: README.md [hunk 5]
```diff
@@ -249,13 +249,13 @@ SET vectors.k = 32;
 SET LOCAL vectors.k = 32;
 ```
 
+
+### Filter in vector search
+
 If you want to disable vector indexing or prefilter, we also offer some GUC options:
 - `vectors.enable_vector_index`: Enable or disable the vector index. Default value is `on`.
 - `vectors.enable_prefilter`: Enable or disable the prefilter. Default value is `on`.
```
Comment (usamoi) [inline README.md:257]: It should be `off` in current codebase.

File: README.md [hunk 6]
```diff
@@ -249,13 +249,13 @@ SET vectors.k = 32;
...
-
-## Limitations
-
-- The filtering process is not yet optimized. To achieve optimal performance, you may need to manually experiment with different strategies. For example, you can try searching without a vector index or implementing post-filtering techniques like the following query: `select * from (select * from items ORDER BY embedding <-> '[3,2,1]' LIMIT 100 ) where category = 1`. This involves using approximate nearest neighbor (ANN) search to obtain enough results and then applying filtering afterwards.
+More detailed doc at [./docs/filter.md] and issue [#113](https://github.com/tensorchord/pgvecto.rs/issues/113).
 
 ## Setting up the development environment
```
Comment (usamoi) [inline README.md:260]: It seems that nobody use envd to develop this repo now. Will we afford extra efforts to maintain it?

File: docs/filter.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+The actual performance of filtering depends on how tight your condition is:
+- If your filter is loose, let's say 90% data satisfy it. Performing an ANN search first and then applying a filter to the results would yield the fastest and relatively good outcomes. This is post filtering
+- If your filter is really tight, let's say only 100 rows satisfy it. The optimal approach is to apply the filter first, obtaining 100 results, and then calculate the distance directly without utilizing any vector index. This is brute force
+- If your filter is kind of tight, let's say 20% data satisfy it. The post-filtering strategy may encounter issues as the ANN search might not yield sufficient results for filtering. The optimal approach in this case is prefiltering. As the algorithm traverses the hnsw graph to discover new points, it will simultaneously verify the filter until there are an adequate number of candidates. All the results from the 
+vector index have already met the filter criteria.
+
+To select different mode:
+Prefiltering (default mode): SET vectors.enable_vector_index=on; SET vectors.enable_prefilter=on
```
Comment (usamoi) [inline docs/filter.md:12]: Wrap it in code block.

File: README.md (hunk 6 above)
Comment (VoVAllen) [reply README.md:260]: Removed

File: README.md (hunk 2 above)
Comment (VoVAllen) [reply README.md:17]: fast indexing is comparing to pgvector's single thread indexing

File: README.md (hunk 3 above)
Comment (VoVAllen) [reply README.md:30]: Fixed

File: README.md (hunk 2 above)
Comment (usamoi) [reply README.md:17]: > not only at index creation, but also at insertion with an index

it is confusing

Comment (VoVAllen) [reply README.md:17]: Do you have any suggestion?

Comment (usamoi) [reply README.md:17]: > it is fast

File: README.md [hunk 8]
```diff
@@ -14,23 +14,22 @@ pgvecto.rs is a Postgres extension that provides vector similarity search functi
 ## Why use pgvecto.rs
 
 - 💃 **Easy to use**: pgvecto.rs is a Postgres extension, which means that you can use it directly within your existing database. This makes it easy to integrate into your existing workflows and applications.
+- 🔗 **Async indexing**: pgvecto.rs's index is asynchronously constructed by the background threads. The search is performed by brute force before the async construction is completed. More details can be found [here](./docs/indexing.md)
+- 🌠 **Fast insertion**: New rows are not blocked by the index construction, with the help of async indexing. Enjoy your search immediately!
```
Comment (usamoi) [inline README.md:18]: I think "Async indexing" and "Fast insertion" are the same thing.

File: README.md [hunk 9]
```diff
@@ -240,6 +224,26 @@ Options for table `product`.
 | sample | integer | Samples to be used for quantization. Default value is `65535`.                                                           |
 | ratio  | string  | Compression ratio for quantization. Only `"x4"`, `"x8"`, `"x16"`, `"x32"`, `"x64"` are allowed. Default value is `"x4"`. |
 
+<details>
+<summary>[Advanced] Segment and Optimization Settings</summary>
```
Comment (usamoi) [inline README.md:228]: Why put these paragraphs to a `<details>`?

**Comment by VoVAllen** [comment]
Overall looks good to me. Please include the introduction for different filter choice, based on my comment in https://github.com/tensorchord/pgvecto.rs/issues/113.

File: README.md [hunk 10]
```diff
@@ -14,257 +14,32 @@ pgvecto.rs is a Postgres extension that provides vector similarity search functi
...
-| WAL amplification                           | 2x 😃                              | 30x 🧐                     |
+- [Installation](./docs/installation.md)
+- [Get Started](./docs/get-started.md)
+- [Indexing](./docs/indexing.md)
+- [Searching](./docs/searching.md)
+- [Comparison with pgvector](./docs/comparison-pgvector.md)
+- [Why not a specialty vector database?](./docs/comparison-with-specialized-vectordb.md)
 
-And based on our benchmark, pgvecto.rs can be up to 2x faster than pgvector on hnsw indexes with same configurations. Read more about the comparison at [here](./docs/comparison-pgvector.md).
+For users, we recommend you to try pgvecto.rs using our pre-built docker, by running
```
Comment (gaocegege) [inline README.md:31]:
```suggestion
For users, we recommend you to try pgvecto.rs using our pre-built docker image, by running
```

File: README.md [hunk 11]
```diff
@@ -14,257 +14,32 @@ pgvecto.rs is a Postgres extension that provides vector similarity search functi
 ## Why use pgvecto.rs
 
 - 💃 **Easy to use**: pgvecto.rs is a Postgres extension, which means that you can use it directly within your existing database. This makes it easy to integrate into your existing workflows and applications.
+- 🔗 **Async indexing**: pgvecto.rs's index is asynchronously constructed by the background threads and does not block insertions and always ready for new queries.
```
Comment (gaocegege) [inline README.md:17]: Could we add some numbers to show the benefit?

File: docs/comparison-pgvector.md [hunk 12]
```diff
@@ -1,5 +1,15 @@
...
+|                                             | pgvecto.rs                                             | pgvector                |
+| ------------------------------------------- | ------------------------------------------------------ | ----------------------- |
+| Transaction support                         | ✅                                                      | ⚠️                       |
+| Sufficient Result with Delete/Update/Filter | ✅                                                      | ⚠️                       |
+| Vector Dimension Limit                      | 65535                                                  | 2000                    |
+| Prefilter on HNSW                           | ✅                                                      | ❌                       |
+| Parallel Index build                        | ⚡️ Linearly faster with more cores                      | 🐌 Only single core used |
+| Async Index build                           | Ready for queries anytime and do not block insertions. | ❌                       |
+| Quantization                                | Scalar/Product Quantization                            | ❌                       |
+
```
Comment (gaocegege) [inline docs/comparison-pgvector.md:12]: Why dont we keep it in README?

Comment (usamoi) [reply docs/comparison-pgvector.md:12]: README is too long before.

File: README.md [hunk 13]
```diff
@@ -14,257 +14,32 @@ pgvecto.rs is a Postgres extension that provides vector similarity search functi
...
-
-And you can change the number of expected result (such as `ef_search` in hnsw) by using the following SQL.
-
-```sql
----  (Optional) Expected number of candidates returned by index
-SET vectors.k = 32;
---- Or use local to set the value for the current session
-SET LOCAL vectors.k = 32;
+```sh
+docker run --name pgvecto-rs-demo -e POSTGRES_PASSWORD=mysecretpassword -p 5432:5432 -d tensorchord/pgvecto-rs:pg16-latest
```
Comment (gaocegege) [inline README.md:34]: We dont have such a tag.

https://hub.docker.com/repository/docker/tensorchord/pgvecto-rs/tags?page=1&ordering=last_updated&name=pg16-latest

Comment (usamoi) [reply README.md:34]: it should be fixed in #159

File: README.md (hunk 11 above)
Comment (VoVAllen) [reply README.md:17]: we'll have a seperate PR for all the benchmarks including insert/index/query performance

File: docs/comparison-pgvector.md (hunk 12 above)
Comment (VoVAllen) [reply docs/comparison-pgvector.md:12]: I think we can keep this table in readme?

Comment (usamoi) [reply docs/comparison-pgvector.md]: moved back

**Pull request merged**